### PR TITLE
Feature/issue#294 Account

### PR DESCRIPTION
**Pull request by alexbojko**
Includes issue#288, issue#293, issue#292, issue#290, issue#291, issue#289, issue#290 
(signup, login, reset password, email confirmation, change password)

### CODE REVIEW

File: mysite/psa/forms.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            raise forms.ValidationError(
+                "Fields {} and {} should have the same values".format(
+                    'email',
+                    'email confirmation')
+            )
+
+    def clean_email(self):
+        email = self.cleaned_data['email']
+        already_exists_exc = forms.ValidationError('This email is already registered in the system.')
+        if User.objects.filter(email=email):
```
Comment (maxsocl) [inline mysite/psa/forms.py]: @alexbojko change pls:
```
if User.objects.filter(email=email) or CustomCode.objects.filter(email=email, verified=True):
    raise already_exists_exc
```

File: mysite/psa/forms.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class SocialForm(forms.ModelForm):
+    class Meta:
+        model = Instructor
+        fields = (
+            'user',
+            'institution',
+            'id',
+            # 'department',
```
Comment (maxsocl) [inline mysite/psa/forms.py]: @alexbojko remove commented out code pls

File: mysite/psa/views.py [hunk 3]
```diff
@@ -52,30 +57,119 @@ def validation_sent(request):
...
+            password=password,
+            first_name=kwargs['first_name'],
+            last_name=kwargs['last_name'],
+
+        )
+    else:
+        username += '_'
+        return check_username_and_create_user(username, email, password, **kwargs)
+
+def send_confirmation_email(user):
```
Comment (maxsocl) [inline mysite/psa/views.py]: @alexbojko are you using this func now?

File: mysite/psa/views.py [hunk 4]
```diff
@@ -52,30 +57,119 @@ def validation_sent(request):
...
+    '''
+    This function will send email about successful registration to user.
+    :param user: user instance
+    :return: None
+    '''
+    return True
+
+def signup(request, next_page=None):
+    """
+    Fields to handle on:
```
Comment (maxsocl) [inline mysite/psa/views.py]: @alexbojko change docstring according to comment above

File: mysite/psa/views.py [hunk 5]
```diff
@@ -52,30 +57,119 @@ def validation_sent(request):
...
+            username = form.cleaned_data['email'].split('@', 2)[0]
+            user = check_username_and_create_user(
+                username=username,
+                **form.cleaned_data
+            )
+            instructor = Instructor.objects.create(
+                user=user,
+                institution=form.cleaned_data['institution'],
+            )
+            # params = form.cleaned_data
```
Comment (maxsocl) [inline mysite/psa/views.py]: @alexbojko remove all unused code (commented out)

File: mysite/psa/views.py [hunk 6]
```diff
@@ -52,30 +57,119 @@ def validation_sent(request):
...
+            # username = user.username
+            # password = params['password']
+
+            request.user = user
```
Comment (maxsocl) [inline mysite/psa/views.py]: @alexbojko add comment to describe this trick

File: mysite/psa/views.py [hunk 7]
```diff
@@ -52,30 +57,119 @@ def validation_sent(request):
...
+            response = complete(request, 'email')
+            request.user = AnonymousUser()
+            return response
+            # user = authenticate(username=username, password=password)
```
Comment (maxsocl) [inline mysite/psa/views.py]: @alexbojko remove all unused code (commented out)

File: mysite/templates/accounts/settings.html [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  <h4>Please confirm your email to be able to restore progress.</h4>
+  <h4>This session may expire without email confirmation.</h4>
+  <h4>Send email confirmation link one more time <a class="btn btn-default" name="email-required-modal_2" href="{% url "social:begin" "email" %}">
+  <i class="fa fa-envelope"></i>
+  re-send
+  </a>
+  </h4>
+  {% endif %}
+
+{% comment %}
```
Comment (maxsocl) [inline mysite/templates/accounts/settings.html]: @alexbojko Why didn't you simply remove this code?

File: mysite/accounts/views.py [hunk 9]
```diff
@@ -1,3 +1,79 @@
...
+        return render(
+            request,
+            'accounts/settings.html',
+            dict(
+                user_form=UserForm(instance=request.user),
+                instructor_form=InstructorForm(instance=request.user.instructor),
+                password_form=ChangePasswordForm(),
+                delete_account_form=DeleteAccountForm(instance=request.user),
+                email_form=ChangeEmailForm(initial={'email': request.user.email}),
+                person=request.user
```
Comment (maxsocl) [inline mysite/accounts/views.py]: @alexbojko need to add Mixin to chech for `anonymous` username.

File: mysite/templates/accounts/settings.html [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+{% block title %}
+Account settings page
+{% endblock %}
+
+{% block content %}
+{% if person and user.id == person.id %}
+
+<div class="buttons">
+  {% if user.is_authenticated and not 'anonymous' in user.username %}
```
Comment (maxsocl) [inline mysite/templates/accounts/settings.html]: @alexbojko this page should not be rendered for users w/  `anonymous` username.

File: mysite/templates/accounts/settings.html [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    <div class="user-details">
+      You are logged in as <code>{{ user.username }}</code>!
+    </div>
+  {% endif %}
+  {% if user and 'anonymous' in user.username %}
```
Comment (maxsocl) [inline mysite/templates/accounts/settings.html]: @alexbojko remove this pls

File: mysite/templates/accounts/settings.html [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+  <div class="legacy">
+    {% include 'psa/legacy_backends.html' %}
+  </div>
+  {% endif %}
+{% endcomment %}
+
+<div class="container-fluid UserForm" style="text-align: left">
+{% if user and not 'anonymous' in user.username %}
+  {% comment %}
```
Comment (maxsocl) [inline mysite/templates/accounts/settings.html]: @alexbojko remove commented out code

File: mysite/templates/accounts/settings.html [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+  </a>
+
+  <div class="legacy">
+    {% include 'psa/legacy_backends.html' %}
+  </div>
+  {% endif %}
+{% endcomment %}
+
+<div class="container-fluid UserForm" style="text-align: left">
+{% if user and not 'anonymous' in user.username %}
```
Comment (maxsocl) [inline mysite/templates/accounts/settings.html]: @alexbojko remove check for `anonymous` in username

File: .travis.yml [hunk 14]
```diff
@@ -10,6 +10,7 @@ addons:
 install:
   - pip install -r requirements.txt -U pip
 before_script:
+  - mv mysite/mysite/settings/local_example.py mysite/mysite/settings/local.py
```
Comment (maxsocl) [inline .travis.yml]: @alexbojko this is wrong commit in this PR

File: mysite/templates/psa/signup.html [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+      </div>
+      {% else %}
+      {{ field }}
+      {% endif %}
+    {% endfor %}
+      {% if form.errors %}
+      {% comment %} {{ form.errors }} {% endcomment %}
+      {% endif %}
+
+    {% comment %} {%if next %}
```
Comment (maxsocl) [inline mysite/templates/psa/signup.html]: @alexbojko remove commented out code

File: mysite/templates/psa/signup.html [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+        <input type="hidden" name="next" value="{{ next }}" />
+    {% endif %}
+    {% endcomment %}
+    <br>
+    <div class="control-group">
+        <div class="controls">
+            <button type="submit" class="btn">Login</button>
```
Comment (maxsocl) [inline mysite/templates/psa/signup.html]: @alexbojko change Button name to Register

File: mysite/templates/psa/signup.html [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+<br>
+<div class="container-fluid">
+{% if not user.is_authenticated %}
+{% endif %}
+  {% include "psa/legacy_backends.html" %}
+</div>
+<br>
+
+<div id="login">
```
Comment (maxsocl) [inline mysite/templates/psa/signup.html]: @alexbojko Why do you use `login` for Register form?

File: mysite/templates/psa/signup.html [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+<h4>Signup:</h4>
+    <form class="form-horizontal" name="LoginForm" action="." method="post">
```
Comment (maxsocl) [inline mysite/templates/psa/signup.html]: @alexbojko change form name to something w/ resigter name

File: mysite/templates/psa/new_custom_login.html [hunk 19]
```diff
@@ -0,0 +1,46 @@
+{% extends "ct/portal.html" %}
```
Comment (maxsocl) [inline mysite/templates/psa/new_custom_login.html]: @alexbojko our new templates should extend something new but not portal template.

File: mysite/templates/registration/password_change_done.html [hunk 20]
```diff
@@ -0,0 +1,8 @@
+{% extends "ct/portal.html" %}
```
Comment (maxsocl) [inline mysite/templates/registration/password_change_done.html]: @alexbojko our new templates should extend something new but not portal template.

File: mysite/templates/registration/password_change_form.html [hunk 21]
```diff
@@ -0,0 +1,53 @@
+{% extends "ct/portal.html" %}
```
Comment (maxsocl) [inline mysite/templates/registration/password_change_form.html]: @alexbojko our new templates should extend something new but not portal template.

File: mysite/templates/registration/password_reset_complete.html [hunk 22]
```diff
@@ -0,0 +1,13 @@
+{% extends "ct/portal.html" %}
```
Comment (maxsocl) [inline mysite/templates/registration/password_reset_complete.html]: @alexbojko same as above

File: mysite/templates/registration/password_reset_confirm.html [hunk 23]
```diff
@@ -0,0 +1,25 @@
+{% extends "ct/portal.html" %}
```
Comment (maxsocl) [inline mysite/templates/registration/password_reset_confirm.html]: @alexbojko same as above

File: mysite/templates/registration/password_reset_done.html [hunk 24]
```diff
@@ -0,0 +1,19 @@
+{% extends "ct/portal.html" %}
```
Comment (maxsocl) [inline mysite/templates/registration/password_reset_done.html]: @alexbojko same as above

File: mysite/templates/registration/password_reset_done.html [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+{% load i18n %}
+
+{% block breadcrumbs %}
+<div class="breadcrumbs">
+<a href="{% url 'admin:index' %}">{% trans 'Home' %}</a>
```
Comment (maxsocl) [inline mysite/templates/registration/password_reset_done.html]: @alexbojko why do we need this?

File: mysite/templates/registration/password_reset_form.html [hunk 26]
```diff
@@ -0,0 +1,22 @@
+{% extends "ct/portal.html" %}
```
Comment (maxsocl) [inline mysite/templates/registration/password_reset_form.html]: @alexbojko our new templates should extend something new but not portal template.

File: mysite/templates/registration/password_reset_form.html [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+{% load i18n %}
+
+{% block breadcrumbs %}
+<div class="breadcrumbs">
+<a href="{% url 'admin:index' %}">{% trans 'Home' %}</a>
```
Comment (maxsocl) [inline mysite/templates/registration/password_reset_form.html]: @alexbojko why do we need this?

File: mysite/accounts/views.py (hunk 9 above)
Comment (alexbojko) [inline mysite/accounts/views.py]: it's already present there

Comment (maxsocl) [inline mysite/accounts/views.py]: @alexbojko Where?

Comment (maxsocl) [inline mysite/accounts/views.py]: @alexbojko Great! Can you can some tests for this case?

**Pull request merged**